[WIP] Turn off notifications for posts' reactions

nicolas-amabile opened the pull request:
<!--- Prepend PR title with [WIP] if work in progress. Remove when ready for review. -->

## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Documentation Update

## Description
This PR adds the feature of disabling notifications for post reactions.

## Related Tickets & Documents
Feature request here: https://github.com/thepracticaldev/dev.to/issues/3244

## Mobile & Desktop Screenshots/Recordings (if there are UI changes)
![demo](https://user-images.githubusercontent.com/8633440/60762637-941db400-a03b-11e9-86bc-f8afc660f826.gif)

## Added to documentation?

- [ ] docs.dev.to
- [ ] readme
- [x] no documentation needed

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/thepracticaldev/dev.to?pullRequest=3391) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/thepracticaldev/dev.to?pullRequest=3391) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/thepracticaldev/dev.to?pullRequest=3391) it.</sub>

rhymes commented:
Hi @nicolas-amabile, thanks for the PR!  I'm going to leave my feedback there are few things missing / to discuss, I always encourage people to start a discussion on new features on the issue, in this case #3244, than to start coding right away. This way both the team and the contributor knows how and what.

In this case I think your PR is incomplete:

* the receiver can be an organization, yet you're adding the field only to the user table
* `receiver_data` is a Hash with two fields: the class (User or Organization) and the ID, which means that `receiver_data.fetch(:post_reaction_notifications, true)` it's always going to return true.
* the migration that adds the field `post_reaction_notifications` is missing
* the UI for the organization needs a little bit of thinking
* there are no tests

If it's otherwise incomplete because you're working on it, please mark it as `[WIP]` so that it will be reviewed only at the end.

My suggestion is to start a conversation on the issue itself at least about the UI and then go back to the code from there.

nicolas-amabile commented:
@rhymes thank you for the feedback!
I misinterpreted the `status: approved` tag, my bad ✋I just added the [WIP] and I'll comment on the issue.

> the receiver can be an organization, yet you're adding the field only to the user table

Yes, it makes sense to have the same feature for both, users and organizations. 

> `receiver_data` is a Hash with two fields: the class (User or Organization) and the ID, which means that `receiver_data.fetch(:post_reaction_notifications, true)` it's always going to return true.

Got it, typo when I extracted the logic into the `shoud_send_reaction_notification` function.
It should be `receiver` instead of `receiver_data`.
<img width="778" alt="Screen Shot 2019-07-07 at 12 18 48 PM" src="https://user-images.githubusercontent.com/8633440/60770510-b6521900-a0b1-11e9-9fca-7baa7af490c4.png">
<img width="865" alt="Screen Shot 2019-07-07 at 12 20 13 PM" src="https://user-images.githubusercontent.com/8633440/60770511-b6eaaf80-a0b1-11e9-8765-296d788f3054.png">

> - the migration that adds the field post_reaction_notifications is missing
> - the UI for the organization needs a little bit of thinking
> - there are no tests

Agree, I'll need some help here, I've never used ruby before.

mstruve commented:
Hey @nicolas-amabile!  

Thank you so much for jumping in and trying to fix this issue! With that said, we are in the process of going through and cleaning up some of our PR history to help us better prioritize what we need to be focusing on. Since this PR has not been active for a while and has some conflicts I am going to close it for now. If you would still like to get this merged please feel free to resolve the conflicts, add the additional suggested content, and then open the PR back up. We would be more than happy to help push this through if you would like!

Thanks so much for contributing! Ping us anytime if you have questions!
